File: Leetcode/cpp/sol383.cpp
// https://leetcode.com/problems/ransom-note/description/


class Solution {
public:

    unordered_map<char, int> counter(string str){
        unordered_map<char, int> count;
        for(auto i: str){
            count[i]++;
        }
        return count;
    }

    vector<int> counter_list(string str){
        vector<int> count(26, 0);
        for (int i:str){
            count[i-'a']++;
        }

        return count;
    }

    bool canConstruct(string ransomNote, string magazine) {
        vector<int> note = counter_list(ransomNote);
        vector<int> letters = counter_list(magazine);
        for(int i = 0; i < 26;i++){
            if (!(note[i] <= letters[i])){
                return false;
            }
        }
        return true;

        // for(int i: ransomNote){
        //     int pos = i - 'a';
        //     if (!(note[pos] <= letters[pos])){
        //         return false;
        //     }
        // }


        // unordered_map<char, int> note = counter(ransomNote);
        // unordered_map<char, int> avalaible_letters = counter(magazine);
        // for(int i:temp){
        //     cout << i << ",";
        // }
        // for (auto i:ransomNote){
        //     if(!(note[i]<=avalaible_letters[i])){
        //         return false;
        //     }
        // }
        // return true;
    }
};

File: Leetcode/typescript/sol383.ts

// https://leetcode.com/problems/ransom-note/description/

const counter = (word) => {
    const count = new Map<string, number>();
    for(const ch of word){
        count.set(ch, (count.get(ch) ?? 0)+1)
    }
    return count;
}

function canConstruct(ransomNote: string, magazine: string): boolean {
    const target = counter(ransomNote)
    const source = counter(magazine)
    for(const [ch, freq] of target.entries()){
        if(!source.has(ch) || target.get(ch) > source.get(ch)){
            return false;
        }
    }
    return true;
};